### PR TITLE
Prevent user from adding items with the same name

### PR DESCRIPTION
**Pull request by nur-haziq**
### Main changes
- The `addItem` method in `ItemList` will first check for the existence of an item with the same name in the inventory, before proceeding to add the new item. If an item already exists, the following will be returned:
>An item with the name 'NAME' already exists in the inventory. Please use a different name.
- #211 

### Minor changes
- Added JUnit tests for other methods in `ItemList`
&nbsp;
- Included output statement for when the `search` command returns an empty list
- #217 

---

@PureUsagi I have reverted the behaviour of the `delete` command back to its original behaviour. Do assist in checking this. Thank you!
@nkotaa  Thanks for suggesting the alternative approach.

### CODE REVIEW

File: src/main/java/seedu/binbash/inventory/ItemList.java [hunk 1]
```diff
@@ -311,14 +310,24 @@ private void updateItemThreshold(Item item, int itemThreshold) {
...
-        throw new InvalidCommandException("Item with name '" + itemName + "' not found.");
+
+        if (foundItems.size() > 1) {
+            throw new InvalidCommandException(
+                    String.format(
+                    "There are multiple items containing the string '%s'", itemName)
+                    + System.lineSeparator() + System.lineSeparator()
+                    + "Please provide a more specific name or consider using the item's index, " +
+                    "which can be found using the list command.");
+        }
```
Comment (nkotaa) [inline src/main/java/seedu/binbash/inventory/ItemList.java:328]: While I understand the purpose of this check, this implementation would still make it impossible to say delete "snake plant" when "snake plant seeds" is also in the item list even though the exact non-duplicate name "snake plant" has been specified.

File: src/main/java/seedu/binbash/inventory/SearchAssistant.java [hunk 2]
```diff
@@ -18,7 +18,7 @@ public class SearchAssistant {
     private ArrayList<Item> foundItems;
 
     /**
-     * Note that the list of items to search through has to be explicity set by the caller.
+     * Note that the list of items to search through has to be explicitly set by the caller.
```
Comment (nkotaa) [inline src/main/java/seedu/binbash/inventory/SearchAssistant.java:21]: In light of the above comment I believe a true solution would be a searchByExactName() method in the `SearchAssistant` class.

File: src/test/java/seedu/binbash/inventory/ItemListTest.java [hunk 3]
```diff
@@ -276,4 +292,108 @@ void updateItemDataByName_noChanges_noChangesMade() throws InvalidCommandExcepti
...
+    }
+
+    @Test
+    void sellOrRestockItem_validRestockOperation_success() throws InvalidCommandException {
+        ItemList itemList = new ItemList(new ArrayList<>());
+        itemList.addItem("operational", "Item1", "Description1", 10, LocalDate.MIN, 0.0, 10.0, 5);
+        itemList.sellOrRestockItem("Item1", 5, "restock");
+        OperationalItem item = (OperationalItem) itemList.getItemList().get(0);
+        assertEquals(15, item.getItemQuantity());
+    }
```
Comment (nkotaa) [inline src/test/java/seedu/binbash/inventory/ItemListTest.java:318]:
```suggestion
    void sellOrRestockItem_validRestockOperation_success() {
        try {
            ItemList itemList = new ItemList(new ArrayList<>());
            itemList.addItem("operational", "Item1", "Description1", 10, LocalDate.MIN, 0.0, 10.0, 5);
            itemList.sellOrRestockItem("Item1", 5, "restock");
            OperationalItem item = (OperationalItem) itemList.getItemList().get(0);
            assertEquals(15, item.getItemQuantity());
        } catch (InvalidCommandException e) {
            Assertions.fail();
        }
    }
```

**Review by nkotaa — COMMENTED** [review]
We could also prevent adding items with the same name.

File: src/main/java/seedu/binbash/inventory/SearchAssistant.java (hunk 2 above)
Comment (nur-haziq) [reply src/main/java/seedu/binbash/inventory/SearchAssistant.java:21]: Indeed, this seems like a better solution.

File: src/test/java/seedu/binbash/inventory/ItemListTest.java (hunk 3 above)
Comment (nur-haziq) [reply src/test/java/seedu/binbash/inventory/ItemListTest.java:318]: Implemented. Thanks

**Review by YHWong20 — APPROVED** [review]
Thanks for implementing this feature. Kindly help to update the UG with the necessary info as well!

File: src/main/java/seedu/binbash/command/AddCommand.java [hunk 4]
```diff
@@ -10,30 +9,6 @@
...
-
-                    // Optionally match 'e/' followed by the expiration date.
-                    "(e/(?<itemExpirationDate>.+?)(?=s/))?\\s*" +
-
-                    // Match 's/' followed by the sale price, until seeing 'c/'.
-                    "(s/(?<itemSalePrice>.+?))(?=c/)" +
-
-                    // Finally, match 'c/' followed by the cost price.
-                    "c/(?<itemCostPrice>.+)"
-    );
```
Comment (nkotaa) [inline src/main/java/seedu/binbash/command/AddCommand.java:36]: I don't know how this managed to live past v2.0 so thank you for this.

File: src/main/java/seedu/binbash/inventory/SearchAssistant.java [hunk 5]
```diff
@@ -58,6 +59,23 @@ public SearchAssistant searchByName(String nameField) {
...
+     * @param nameField The exact name to search for in the item names.
+     * @return The current instance of SearchAssistant with the filtered list of items.
+     */
+    public SearchAssistant searchByExactName(String nameField) {
+        if (nameField.equals("")) {
+            return this;
+        }
+        foundItems = foundItems.stream()
+                .filter(item -> item.getItemName().equals(nameField))
+                .collect(Collectors.toCollection(ArrayList::new));
```
Comment (nkotaa) [inline src/main/java/seedu/binbash/inventory/SearchAssistant.java:75]: Before returning, adding an `assert foundItems.size() <= 1` would be nice.

**Review by nkotaa — APPROVED** [review]
Thanks for considering my suggestions.

File: src/main/java/seedu/binbash/command/AddCommand.java (hunk 4 above)
Comment (nur-haziq) [reply src/main/java/seedu/binbash/command/AddCommand.java:36]: Agreed...

**Pull request merged**